src/main/java/seedu/address/logic/commands/MatchPropertyCommand.java

```diff
@@ -46,15 +60,23 @@ public CommandResult execute(Model model) throws CommandException {
 
         Property propertyToMatch = lastShownList.get(targetIndex.getZeroBased());
 
-        // Start constructing the predicate
-        Predicate<Buyer> matchingPredicate = new FilterBuyerByPricePredicate(propertyToMatch.getPrice());
+        ArrayList<Predicate<Buyer>> predicatesList = new ArrayList<>();
+        predicatesList.add(new FilterBuyerByPricePredicate(propertyToMatch.getPrice()));
         if (propertyToMatch.getCharacteristics().isPresent()) {
-            matchingPredicate = matchingPredicate.and(new FilterBuyerByCharacteristicsPredicate(
-                    propertyToMatch.getCharacteristics().get()));
+            predicatesList.add(
+                    new FilterBuyerContainingAnyCharacteristicPredicate(propertyToMatch.getCharacteristics().get()));
+        }
+
+        Optional<Predicate<Buyer>> combinedPredicate;
```

Comment:
Do we need this in an Optional if at least one predicate will be specified?

Reply:
It was needed because the reduce command gives an Optional. But I have refactored a little to make it clearer.

```java
        // predicatesList must not be empty, since at least FilterBuyerByPricePredicate should be added
        assert(!predicatesList.isEmpty());

        Predicate<Buyer> combinedPredicate;
        if (isMatchingAll) {
            combinedPredicate = predicatesList.stream().reduce(Predicate::and).get();
        } else {
            combinedPredicate = predicatesList.stream().reduce(Predicate::or).get();
        }

```